### PR TITLE
Update read.py

### LINKED ISSUE #4: Replace NaN values
### Problem
Our Test data seems to contain numeric substitutes for NaN values. E.g. with one of the testfiles:

```python
from bridget.io import read_TK2_file
df = read_TK2_file('./data/Fendt_TK2_result_2014.csv')
df.CO2.where(df.CO2 < 0).dropna()
```
yields:
image

and the plot of the time series looks like:

image

### Suggestion

We should verify, that this is actually a NaN (@sihassl ?) and check the other files and variables for other values (@AlexDo1 ). Is it always the same?
Then we need to implement this as a default NaN value option to convert it properly to `numpy.NaN`.

### LINKED ISSUE #7: DateTime columns are not of type DateType automatically
When reading in the TK2 files, the DateTime columns (t_begin, t_end, t_mid) are not of type DateTime.

image

The columns need to be adjusted to DateTime by hand (also necessary to plot data over time):

image

---
Should I add this code to the _reader function?

### PR DESCRIPTION
**Pull request by AlexDo1**
Added na_values (probably not the best way, but the only way I could get it running), resulting in the error:
`ValueError: Length mismatch: Expected axis has 54 elements, new values have 61 elements`
when using the read_TK2_file function to read in a csv file (as Mirko expected).

### CODE REVIEW

File: bridget/io/read.py [hunk 1]
```diff
@@ -3,48 +3,50 @@
...
     strict : bool
-        If True, the data structure will be checked to include all columns 
+        If True, the data structure will be checked to include all columns
         expected. On failure a ValueError will be raised.
+    na_values : scalar, str, list-like, or dict, optional
+        Additional values to recognize as NA/NaN.
     """
-    return _reader(fname, 'TK2', strict=strict)
+    read_kwargs = dict(na_values = na_values)
+    return _reader(fname, 'TK2', strict=strict, **read_kwargs)
```
Comment (mmaelicke) [inline bridget/io/read.py]: I would rather pass it as 
```python
return _reader(fname, 'TK2', strict=strict, read_kwargs=read_kwargs)
```
and undo the change below

File: bridget/io/read.py [hunk 2]
```diff
@@ -3,48 +3,50 @@
...
     strict : bool
-        If True, the data structure will be checked to include all columns 
+        If True, the data structure will be checked to include all columns
         expected. On failure a ValueError will be raised.
     """
     return _reader(fname, 'QA_QC', strict=strict)
 
 
-def _reader(fname, type_str, strict=True, read_kwargs={}):
+def _reader(fname, type_str, strict=True, **read_kwargs):
```
Comment (mmaelicke) [inline bridget/io/read.py]: I would keep the old version. See the changes above to make it work.

File: bridget/io/read.py [hunk 3]
```diff
@@ -3,48 +3,50 @@
...
     # read the file
-    data = pd.read_csv(fname, sep=',', header=None, index_col=None, **read_kwargs)
```
Comment (mmaelicke) [inline bridget/io/read.py:46]: Here again, I would keep the old version.
The benefit is, that we can pass all the kwargs we want to further pass to read_csv. That enables us in the future to add another *_kwargs argument to the top-level function.

File: bridget/io/read.py [hunk 4]
```diff
@@ -3,48 +3,50 @@
...
-    
+    data = pd.read_csv(fname, sep=',', header=None, index_col=None, na_values=[read_kwargs['na_values']])
     # drop empty columns
     data.dropna(axis=1, how='all', inplace=True)
```
Comment (mmaelicke) [inline bridget/io/read.py]: Colud you check if we loose the 7 coulmns full of NAN here, or already in the line above?

If we lose them here, we need a workaround. Maybe you can come up with something.

**Review by mmaelicke — CHANGES_REQUESTED** [review]
Looks good. I have some suggestions how we could work on the expected shape mismatch error

**Comment by mmaelicke** [comment]
If you can't make my changes work, just let me know and I will work on it

**Comment by AlexDo1** [comment]
The changes for the usage of kwargs already work.

We are losing the columns with this line of code:
`data.dropna(axis=1, how='all', inplace=True)`
as some columns only consist of the number -9999.9003906 (replaced with NaN before) and so they are considered empty and are droped which leads to a mismatch with the given standard header values.
Commenting out the dropna line leads to the same error as we have an empty last column in the raw data which has to be dropped, otherwise we have a column too much. 

An easy workaround would be to delete the empty last column in the raw data but if this is a problem with all the expected input data files I should code soemthing that does that automatically.

**Comment by mmaelicke** [comment]
Yes, I would prefer the 

>  code something that does that automatically

solution. Have you checked which columns are affected? Maybe we actually have to split them into mandatory and optional default columns.

**Comment by mmaelicke** [comment]
Could work like:

1. Expand the default columns file with that 'NAN column', to make import and column header merging work.

2. Remove the dropna line

3. Explicitly drop the NAN column and import the others with possible NaN values

4. If `strict=True`, drop all columns that have NA and check if the columns array is still correct.

**Comment by AlexDo1** [comment]
Updated read.py and the default column file.
By deleting the na_column at the end of _reader should make `strict=True` work like  before.

File: bridget/io/default_columns_TK2.csv [hunk 5]
```diff
@@ -59,3 +59,4 @@ noise_ustar,noise error for ustar,%
 noise_hts,noise error for HTs,%
 noise_lve,noise error for LvE,%
 noise_co2,noise error for Cov(w'CO2')/NEE,%
+na_column,empty_column,delete_this_column_in_read_function
```
Comment (mmaelicke) [inline bridget/io/default_columns_TK2.csv:62]: I like.

File: bridget/io/read.py [hunk 6]
```diff
@@ -44,16 +47,16 @@ def read_QA_QC_file(fname, strict=True):
...
     # set the column names
     data.columns = get_header(type_str)
-
     # strict mode
     if strict:
         assert (get_header(type_str) == data.columns).all()
-
+    # raw data includes an empty last column, delete it
+    data = data.drop('na_column', axis=1)
+    # convert datetime columns to datetime data type
```
Comment (mmaelicke) [inline bridget/io/read.py:57]: maybe we can add some kind of 
```python
if not df.t_begin.dtype in (datetime64, datetime):
  df['t_begin'] =  [...]
```
where `datetime` is the `datetime` class from the module of same name and the other one is from numpy I think
It's not super important, so you can accept the PR also without this fix. I think it's just better if the types are parsed correctly coming from another source

**Review by mmaelicke — APPROVED** [review]
Looks good. I'll approve, although there is a minor thing that could be implemented

File: bridget/io/read.py (hunk 6 above)
Comment (AlexDo1) [reply bridget/io/read.py:57]: I'm just trying to implement that. The problem ist that the columns of interest seem to be of the type datetime.datetime (as far as I understood it, datetime.datetime just looks if the input could be a date) so the if statement is not working.
But it is necessary to convert the columns with
`data['t_begin'] =  pd.to_datetime(data['t_begin'], format='%d.%m.%Y %H:%M')`
to plot the data.

Comment (mmaelicke) [reply bridget/io/read.py:57]: `datetime.datetime` is the built-in date library in python. But as far as I understood it, the usage of `to_datetime`, with given format string converts `str` columns to datetime columns.

We could also always run that line, accept the PR and open another issue for implementing the if-clause.

Comment (mmaelicke) [reply bridget/io/read.py:57]: I we will definitely have to implement some unittests as soon as we have the target data structure defined. But that's on me

**Pull request merged**